File: playground/src/validator.ts
// Browser-side AgentCard validator.
//
// Mirrors what `a2a-testbed validate` does in Python: for each entry
// in `capabilities.extensions[]`, fetch `<uri>/manifest.json`, parse
// the contained JSON Schema, and validate the entry's `params`
// payload. No backend required — works against any manifest hosted at
// a CORS-friendly URL (GitHub Pages serves the headers we need).

import Ajv, { type ErrorObject } from 'ajv';
import addFormats from 'ajv-formats';

export type FindingKind =
  | 'declared_ok'
  | 'declared_invalid'
  | 'manifest_unreachable'
  | 'manifest_malformed'
  | 'no_payload';

export interface Finding {
  uri: string;
  kind: FindingKind;
  detail: string;
  errors?: ErrorObject[];
  manifestName?: string;
  manifestVersion?: string;
}

interface AgentCardLike {
  capabilities?: {
    extensions?: Array<{
      uri: string;
      params?: unknown;
      payload?: unknown;
      [key: string]: unknown;
    }>;
  };
}

// Minimal ExtensionManifest envelope shape we read from the wire.
// Mirrors a2a_testbed.manifest.types.ExtensionManifest; only the
// fields we actually use are typed here.
interface ManifestEnvelope {
  manifest_version: string;
  extension: {
    uri: string;
    name: string;
    version: string;
  };
  agent_card_payload_schema?: Record<string, unknown>;
}

const ajv = new Ajv({
  strict: false, // protocol manifests use vendor keywords; don't fail on them
  allErrors: true,
  coerceTypes: false,
});
addFormats(ajv);

// Cache fetched manifests so re-validating the same card doesn't
// re-fetch the same URL.
const manifestCache = new Map<string, ManifestEnvelope>();

function manifestUrlForUri(uri: string): string {
  return uri.endsWith('/') ? `${uri}manifest.json` : `${uri}/manifest.json`;
}

async function loadManifest(uri: string): Promise<ManifestEnvelope> {
  const cached = manifestCache.get(uri);
  if (cached) return cached;

  const url = manifestUrlForUri(uri);
  const res = await fetch(url, {
    headers: { Accept: 'application/json' },
  });
  if (!res.ok) {
    throw new Error(`HTTP ${res.status} ${res.statusText} for ${url}`);
  }
  const json = (await res.json()) as ManifestEnvelope;
  if (!json.extension || !json.extension.uri) {
    throw new Error('manifest missing required `extension.uri` field');
  }
  manifestCache.set(uri, json);
  return json;
}

function extractParams(entry: { params?: unknown; payload?: unknown }): unknown {
  if (entry.params !== undefined) return entry.params;
  if (entry.payload !== undefined) return entry.payload;
  // Tolerate the convention where params live as siblings of `uri`.
  const reserved = new Set(['uri', 'description', 'required']);
  const extras: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(entry)) {
    if (!reserved.has(k)) extras[k] = v;
  }
  return Object.keys(extras).length ? extras : undefined;
}

export async function validateAgentCard(
  card: AgentCardLike
): Promise<Finding[]> {
  const extensions = card.capabilities?.extensions ?? [];
  if (extensions.length === 0) {
    return [];
  }

  const findings: Finding[] = [];

  for (const entry of extensions) {
    const uri = entry.uri;
    if (!uri) {
      findings.push({
        uri: '(missing)',
        kind: 'manifest_malformed',
        detail: 'extension entry has no uri field',
      });
      continue;
    }

    let manifest: ManifestEnvelope;
    try {
      manifest = await loadManifest(uri);
    } catch (err) {
      findings.push({
        uri,
        kind: 'manifest_unreachable',
        detail: `could not fetch manifest: ${(err as Error).message}`,
      });
      continue;
    }

    const params = extractParams(entry);
    if (params === undefined) {
      findings.push({
        uri,
        kind: 'no_payload',
        detail: 'extension entry has no params payload to validate',
        manifestName: manifest.extension.name,
        manifestVersion: manifest.extension.version,
      });
      continue;
    }

    const schema = manifest.agent_card_payload_schema;
    if (!schema) {
      findings.push({
        uri,
        kind: 'declared_ok',
        detail:
          'manifest declares no payload schema; entry treated as opaque',
        manifestName: manifest.extension.name,
        manifestVersion: manifest.extension.version,
      });
      continue;
    }

    const validate = ajv.compile(schema);
    const ok = validate(params);
    if (ok) {
      findings.push({
        uri,
        kind: 'declared_ok',
        detail: `payload conforms to ${manifest.extension.name} v${manifest.extension.version}`,
        manifestName: manifest.extension.name,
        manifestVersion: manifest.extension.version,
      });
    } else {
      findings.push({
        uri,
        kind: 'declared_invalid',
        detail: 'payload failed JSON Schema validation',
        errors: validate.errors ?? [],
        manifestName: manifest.extension.name,
        manifestVersion: manifest.extension.version,
      });
    }
  }

  return findings;
}

File: src/a2a_testbed/manifest/validator.py
# Copyright 2026 Ravi Kiran Kadaboina
# Licensed under the Apache License, Version 2.0.

"""Validate AgentCards against fetched ExtensionManifests.

This is the universal validator: given a manifest store, the validator
walks the ``capabilities.extensions[]`` list, fetches each manifest,
and validates the entry's payload against the manifest's
``agent_card_payload_schema``. No protocol-specific code. Adding a new
protocol that publishes a manifest costs zero validator changes.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Optional

from jsonschema import Draft202012Validator
from jsonschema.exceptions import ValidationError

from a2a_testbed.manifest.store import ManifestStore


class FindingKind(str, Enum):
    """Outcome of probing one capabilities.extensions[] entry.

    ``DECLARED_OK`` is the only "passed" state. The other values report
    different failure or information modes — together they let a
    third-party reviewer answer "what's wrong with this AgentCard?"
    without having any prior knowledge of the protocols involved.
    """

    DECLARED_OK = "declared_ok"
    DECLARED_INVALID = "declared_invalid"
    MANIFEST_NOT_FOUND = "manifest_not_found"
    MANIFEST_VERSION_UNSUPPORTED = "manifest_version_unsupported"
    PAYLOAD_MISSING = "payload_missing"


@dataclass(frozen=True)
class Finding:
    """One result from validating one declared extension."""

    kind: FindingKind
    extension_uri: str
    detail: str = ""
    schema_errors: tuple[str, ...] = ()


def _extract_payload(entry: dict) -> Optional[dict]:
    """Pull the typed payload out of an extension entry.

    A2A 1.0 doesn't strictly mandate where extension-specific fields
    live; common conventions in observed implementations are
    ``params`` and ``payload``, with some early implementations using
    inline keys on the entry itself.
    """
    if not isinstance(entry, dict):
        return None
    for key in ("params", "payload"):
        candidate = entry.get(key)
        if isinstance(candidate, dict):
            return candidate
    reserved = {"uri", "description", "required"}
    extra = {k: v for k, v in entry.items() if k not in reserved}
    return extra or None


# Manifest envelope versions this validator understands. Bumped only
# when a NEW major manifest_version requires code changes.
SUPPORTED_MANIFEST_MAJOR_VERSIONS = {"1"}


def _major(version: str) -> str:
    return version.split(".", 1)[0]


async def validate_agent_card(
    card: dict,
    *,
    store: ManifestStore,
) -> list[Finding]:
    """Validate every declared extension in an AgentCard against its manifest.

    Returns one Finding per ``capabilities.extensions[]`` entry. Order
    preserved so callers can correlate findings with the source entry.
    """
    findings: list[Finding] = []

    capabilities = card.get("capabilities") if isinstance(card, dict) else None
    if not isinstance(capabilities, dict):
        return findings
    extensions = capabilities.get("extensions") or []
    if not isinstance(extensions, list):
        return findings

    for entry in extensions:
        if not isinstance(entry, dict):
            continue
        uri = entry.get("uri")
        if not isinstance(uri, str) or not uri:
            continue

        manifest = await store.fetch(uri)
        if manifest is None:
            findings.append(
                Finding(
                    kind=FindingKind.MANIFEST_NOT_FOUND,
                    extension_uri=uri,
                    detail=("no manifest available for this URI; cannot validate payload"),
                )
            )
            continue

        if _major(manifest.manifest_version) not in SUPPORTED_MANIFEST_MAJOR_VERSIONS:
            findings.append(
                Finding(
                    kind=FindingKind.MANIFEST_VERSION_UNSUPPORTED,
                    extension_uri=uri,
                    detail=(
                        f"manifest_version={manifest.manifest_version!r} is "
                        "outside the supported set "
                        f"{sorted(SUPPORTED_MANIFEST_MAJOR_VERSIONS)}"
                    ),
                )
            )
            continue

        payload = _extract_payload(entry)
        if payload is None:
            findings.append(
                Finding(
                    kind=FindingKind.PAYLOAD_MISSING,
                    extension_uri=uri,
                    detail=(
                        "extension declared but no payload present "
                        "(checked 'params', 'payload', and inline keys)"
                    ),
                )
            )
            continue

        validator = Draft202012Validator(manifest.agent_card_payload_schema)
        errors = sorted(validator.iter_errors(payload), key=lambda e: e.path)
        if errors:
            findings.append(
                Finding(
                    kind=FindingKind.DECLARED_INVALID,
                    extension_uri=uri,
                    detail="payload failed JSON Schema validation",
                    schema_errors=tuple(_format_error(e) for e in errors),
                )
            )
            continue

        findings.append(
            Finding(
                kind=FindingKind.DECLARED_OK,
                extension_uri=uri,
                detail=(
                    f"payload conforms to {manifest.extension.name} v{manifest.extension.version}"
                ),
            )
        )

    return findings


def _format_error(err: ValidationError) -> str:
    path = "/".join(str(p) for p in err.absolute_path) or "<root>"
    return f"{path}: {err.message}"
